src/Chapter3.hs

```diff
@@ -479,6 +488,35 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data House = H1 | H2 | H3 | H4
+
+data Entertainment = Church | Library
+
+newtype Castle = Castle String
+
+type Wall = ()
+
+data City = City 
+  (Maybe (Maybe Wall, Castle))
+  Entertainment
+  [House]
+
+population :: House -> Int
+population H1 = 1
+population H2 = 2
+population H3 = 3
+population H4 = 4
+
+buildCastle :: String -> City -> City
+buildCastle nn (City c e h) = City (fmap (const $ Castle nn) <$> c) e h
```

Comment:
Although this is correct, I would not recommend using such complicated set of function applications as it makes code difficult to read in terms future perspective, and harder to get new people onboard with the projects 🙂